### ISSUE TITLE
Flutter didReceiveRemoteNotification Not called

### ISSUE THREAD

**Issue by Bathant** [flutter/flutter#52895]
I've created a plugin which support Push Notification and i'm trying to swizzle the App Delegate functions but `application:didReceiveRemoteNotification:fetchCompletionHandler: `not fired 
The notification is received when i do the following:

- swizzle it like this `application:didReceiveRemoteNotification:`
- Set `content_available` to true in the notification payload

I'm just wondering why this is happening, can anyone help me to figure out how can I make it work with `application:didReceiveRemoteNotification:fetchCompletionHandler:`?

**Comment by iapicca** [comment]
Hi @Bathant 
This platform is not meant for assistance on personal code.
Please see https://flutter.dev/community for resources and asking questions like this,
you may also get some help if you post it on Stack Overflow.

Closing, as this isn't an issue with Flutter itself,
if you disagree please write in the comments and I will reopen it
Thank you

**Comment by Bathant** [comment]
Hi @iapicca  thank you for your response, 
but i think the problem in flutter as i tried to call it `application:didReceiveRemoteNotification:fetchCompletionHandler`: without swizzling and it didn't fire 
it fires only once i changed it with` application:didReceiveRemoteNotification:`

**Comment by iapicca** [comment]
Hi again @Bathant 
can you please provide your `flutter doctor -v` , 
your `flutter run --verbose` ,
your `pubspec.yaml` 
and if possible a reproducible minimal code sample.
Thank you

**Comment by Bathant** [comment]
**flutter doctor -v**
```
[✓] Flutter (Channel stable, v1.12.13+hotfix.7, on Mac OS X 10.15.3 19D76, locale en-EG)
    • Flutter version 1.12.13+hotfix.7 at /Users/macbookpro/Developer/flutter
    • Framework revision 9f5ff2306b (8 weeks ago), 2020-01-26 22:38:26 -0800
    • Engine revision a67792536c
    • Dart version 2.7.0

 
[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/
[Flutter-run.txt](https://github.com/flutter/flutter/files/4360065/Flutter-run.txt)
macbookpro/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • iPhone 8 • EBA9FAC9-26C9-4E30-89ED-65BCB5241BD6 • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-3 (simulator)

! Doctor found issues in 1 category.
```
**flutter run --verbose**
[Flutter-run.txt](https://github.com/flutter/flutter/files/4360069/Flutter-run.txt)

**pubspec.yaml**

[pubspec.txt](https://github.com/flutter/flutter/files/4360087/pubspec.txt)

**Comment by Bathant** [comment]
Kindly check the following sample code, 
you can find the issue existing in AppDelegate. 
https://github.com/Bathant/PushNotificationFlutterExample.git

**Comment by iapicca** [comment]
Hi @Bathant 
are you experiencing the issue on a physical device running `flutter run --release --verbose`?
```bash
[   +2 ms] Notification Registration Error Error Domain=NSCocoaErrorDomain Code=3010 "remote notifications are not supported in the
simulator" UserInfo={NSLocalizedDescription=remote notifications are not supported in the simulator}
```
thank you

**Comment by Bathant** [comment]
Hi @iapicca  I know that i can't run the push notification on a simulator 
so i run that via Xcode with a real device. 
and that because when i use this command ` flutter run --release --verbose` it force me to use a simulator 
Here's the output of this command when i connect to a real device 
[Flutter-run.txt](https://github.com/flutter/flutter/files/4363334/Flutter-run.txt)

Thank you for your support.

**Comment by iapicca** [comment]
Hi @Bathant 
given the error
```bash
[ +318 ms] No supported devices connected.
[  +11 ms] "flutter run" took 2,993ms.
```
the app doesn't seem to be running on the physical device
also the [app](https://github.com/flutter/flutter/issues/52895#issuecomment-601763460) sample you provided appear to be just an `hello world` app
could you provide logs and sample representative of the issue
Thank you

Also, give your [doctor](https://github.com/flutter/flutter/issues/52895#issuecomment-601727881)  you might want to run `flutter doctor --android-licenses`

**Comment by Bathant** [comment]
I just attached a simple runner project without any third-party integrated to reproduce the issue easily. and you can navigate to the **AppDelegate** file to see the issue after receiving the notification, 
you will notice that `application:didReceiveRemoteNotification:fetchCompletionHandler: ` will not fire.
I run the project on a physical device via Xcode because the flutter command does not work with my physical device.

**Comment by karedem** [comment]
hi, Maybe I have the same problem
in a native ios example project , didReceiveRemoteNotification was called;
the same way to a flutter project (ios part), didReceiveRemoteNotification  not called;
Looking forward to your reply, Thank you!
and here is the log of `flutter doctor -v`
[✓] Flutter (Channel stable, v1.17.0, on Mac OS X 10.14.6 18G4032, locale
    zh-Hans-CN)
    • Flutter version 1.17.0 at /Users/ghdev/temp/github/flutter
    • Framework revision e6b34c2b5c (4 周前), 2020-05-02 11:39:18 -0700
    • Engine revision 540786dd51
    • Dart version 2.8.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/ghdev/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/ghdev/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.2.1, Build version 11B500
    • CocoaPods version 1.9.1

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Community Edition (version 2019.3.4)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (1 available)
    • iPhone • 063371b8f0073177f47f4979820b95bb452a00bb • ios • iOS 12.4.5

**Comment by karedem** [comment]
i meet the same question,
real reason is  used another lib : `flutter_local_notifications`
then soled the question after remove it

**Comment by Bathant** [comment]
> i meet the same question,
> real reason is used another lib : `flutter_local_notifications`
> then soled the question after remove it

Can you please clarify more how did you solve this?

**Comment by karedem** [comment]
> > i meet the same question,
> > real reason is used another lib : `flutter_local_notifications`
> > then soled the question after remove it
> 
> Hi @karedem
> Can you please clarify more how did you solve this?

i created a new flutter application, then regist remote Notification service in iOS part,
function `didReceiveRemoteNotification with completionHandler` was called,
i found out the function not called because of another plugin `flutter_local_notifications `,  it 's ok when i remove it
so it's no problem in the application, but  i didn't  test the plugin
that's all, good lock to my bro

**Comment by devSC** [comment]
Me too

**Comment by TahaTesser** [comment]
<details> 
<summary>code sample</summary> 

```dart
import UIKit
import Flutter
import UserNotifications

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
    override func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        GeneratedPluginRegistrant.register(with: self)
        registerForPushNotifications()
        return super.application(application, didFinishLaunchingWithOptions: launchOptions)
    }
    func registerForPushNotifications() {
        if #available(iOS 10.0, *) {
            UNUserNotificationCenter.current()
                .requestAuthorization(options: [.alert, .sound, .badge]) {
                    [weak self] granted, error in
                    
                    print("Permission granted: \(granted)")
                    guard granted else { return }
                    self?.getNotificationSettings()
            }
        } else {
            // Fallback on earlier versions
        }
    }
    func getNotificationSettings() {
        if #available(iOS 10.0, *) {
            UNUserNotificationCenter.current().getNotificationSettings { settings in
                print("Notification settings: \(settings)")
                guard settings.authorizationStatus == .authorized else { return }
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
        } else {
            // Fallback on earlier versions
        }
        
    }
    //MARK:- THIS RUN SUCESSFULLY
    
    override func application(_ application: UIApplication,
                                didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
        print("recieved")
    }

//MARK:- HERE IS THE ISSUE
    override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        print("didn't recieved")
    }
    
    override func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        let tokenParts = deviceToken.map { data in String(format: "%02.2hhx", data) }
        let token = tokenParts.joined()
        print("Device Token: \(token)")
    }
    
    override func application(
        _ application: UIApplication,
        didFailToRegisterForRemoteNotificationsWithError error: Error) {
        print("Failed to register: \(error)")
    }

    
    
}
```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.24.0-4.0.pre.107, on Mac OS X 10.15.7 19H2
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-4.0.pre.107 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 35a94f70e1 (7 hours ago), 2020-10-27 23:47:04 -0700
    • Engine revision 64e6599910
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (5 available)
    • RMX2001 (mobile)       • 192.168.0.104:5555        • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.1
    • macOS (desktop)        • macos                     • darwin-x64     • Mac
      OS X 10.15.7 19H2 darwin-x64
    • Web Server (web)       • web-server                • web-javascript •
      Flutter Tools
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 86.0.4240.111

• No issues found!
```

</details>

**Comment by FrantisekGazo** [comment]
This issue is still reproducible in flutter 2.2.3

**Comment by T-Oner** [comment]
This is a very bad problem, the Flutter app cannot receive CloudKit notifications

**Comment by theanhoo** [comment]
I removed `flutter_local_notifications` but the issue persists.

To repeat what @TahaTesser posted:

The following works and the callback is invoked:
```
override func application(_ application: UIApplication,
                          didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
    print("Received remote notification: \(userInfo)")
}
```

This does not work:
```
override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
    print("Received remote notification: \(userInfo)")
        completionHandler(.newData)
}
```

**Comment by pradeeplobo86** [comment]
The issue is reproducible in flutter 3.3.2 

To repeat what @theanhoo and @TahaTesser  posted:

The following works and the callback is invoked
```
override func application(_ application: UIApplication,
                          didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
    print("Received remote notification: \(userInfo)")
}
```
This does not work:
```
override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
    print("Received remote notification: \(userInfo)")
        completionHandler(.newData)
}
```

<details close>
<summary>Flutter frameworks used in sample project</summary>
<br>
<img width="1429" alt="Screen Shot 2022-09-29 at 8 47 35 AM" src="https://user-images.githubusercontent.com/99631125/193064378-36d811a4-4b86-4e96-b588-456ebefc571a.png">
</details>


<details close>
<summary>flutter doctor -v</summary>
<br>

```[✓] Flutter (Channel stable, 3.3.2, on macOS 12.6 21G115 darwin-arm, locale
    en-CA)
    • Flutter version 3.3.2 on channel stable at
      /Users/pradeeplobo/su-flutter-form/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (2 weeks ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version
    32.1.0-rc1)
    • Android SDK at /Users/pradeeplobo/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A309
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Utilities/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (6 available)
    • Android SDK built for arm64 (mobile) • emulator-5554
      • android-arm64  • Android 10 (API 29) (emulator)
    • iPhone SE (3rd generation) (mobile)  •
      0D3CBE91-5DA1-4F70-BE26-DA8E6254B902 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • iPhone 14 Pro (mobile)               •
      F1EDED83-13B8-4D8A-B804-CCFA82059E6C • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • iPad (9th generation) (mobile)       •
      256EFDE1-1D59-41F0-9DCF-8D2913DB2288 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)                      • macos
      • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)                         • chrome
      • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details >

**Comment by dayoul** [comment]
This issue is still reproducible in flutter version 3.3.8
I'm using pure APNS(not the one from flutter-fire) to fetch silent push, and only this one is called when app is in foreground or background.
(It's deprecated according to the [apple document](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623117-application))
```
override func application(_ application: UIApplication,
                          didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
// This method get's called when getting silent push.
}
```
Below is never called.
```
override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
 // This method is supposed to be called when getting silent push, but never called.
}
```
Strange thing is if I don't inherit ```FlutterAppDelegate``` in my ```AppDelegate``` and instead implement protocol ```UIApplicationDelegate```(just like native iOS app), ```didReceiveRemoteNotification:fetchCompletionHandler``` is called!.
I think there's probably a problem in ```FlutterAppDelegate``` since ```didReceiveRemoteNotification``` is now deprecated and the apple document is recommending to use ```didReceiveRemoteNotification:fetchCompletionHandler```.

Any idea to let ```didReceiveRemoteNotification:fetchCompletionHandler``` to be called on silent pushes?

**Comment by dayoul** [comment]
> The swift function signature is missing the return type `Bool`. The below worked for me. Notice the trailing `-> Bool `
> 
> ```swift
> public func application( _ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any],
>   fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool // HERE
> { 
>         
>         print("received notificaiton in background: \(userInfo)")
> 
>         completionHandler(.noData)
>         return true
>     }
> ```

Hello @Samaritan1011001, I've tried it but it's still not called on flutter version 3.3.8. Could you please provide your AppDelegate code and flutter version? Thanks in advance.

**Comment by Samaritan1011001** [comment]
@dayoul -- Oh sorry. My case is a bit different. I am trying to use the function inside my plugin that inherits `FlutterPlugin`

```swift
public class SwiftPushIosPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.test.notification/push", binaryMessenger: registrar.messenger())
        let instance = SwiftPushIosPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    public func application( _ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any],
      fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool // HERE
    { 
        print("received notificaiton in background: \(userInfo)")
        completionHandler(.noData)
        return true
     }
}
```

**Comment by rlepinski** [comment]
This is breaking Airship integration as well. It looks like the FlutterAppDelegate is overriding respondsToSelector methods:
```
- (BOOL)respondsToSelector:(SEL)selector {
  if ([_lifeCycleDelegate isSelectorAddedDynamically:selector]) {
    return [self delegateRespondsSelectorToPlugins:selector];
  }
  return [super respondsToSelector:selector];
}
```

And will not respond to didReceiveRemoteNotificaiton unless a plugin implements it, such as in @Samaritan1011001 example. 

Not sure the best fix, I will leave that up to the Flutter devs, but it should support dynamically adding methods through swizzling and plugins delegates.

**Comment by rlepinski** [comment]
Actually, if the plugin swizzles right then everything will work. The respondsToSelector just needs to be updated to check both locations:

```
- (BOOL)respondsToSelector:(SEL)selector {
  if ([_lifeCycleDelegate isSelectorAddedDynamically:selector] && [self delegateRespondsSelectorToPlugins:selector]) {
    return YES;
  }
  
  return [super respondsToSelector:selector];
}
```

**Comment by darshankawar** [comment]
@Bathant Revisiting this issue to check if it still persists or not, I tried to use your repro but was unable to properly run it using latest stable version. Also, I am not sure how exactly to trigger `didReceiveRemoteNotification` without any plugin (for ex: firebase_messaging). 

Please check below links and see if they help to resolve the issue or not, since there have been lot of updates done to the framework since this issue was filed.


https://firebase.google.com/docs/cloud-messaging/flutter/client#ios

https://github.com/firebase/flutterfire/issues/9902
https://github.com/firebase/flutterfire/issues/1683

If none of above helps and issue still persists using latest version, please provide us updated repo and steps to replicate so that we can properly address the issue.

**Comment by darshankawar** [comment]
Closing this based on no activity from OP and others who had this issue and also considering age of the issue at the time it was filed. If anybody thinks this is stil an issue using latest stable version, please file new issue using template so that we can properly address it.
